### PR TITLE
Update ChemicalEnvironments SMIRKS parsing

### PR DESCRIPTION
**Pull request by bannanc**
I have updated ChemicalEnvironments to parse more SMIRKS, per issue #106. Please note, I pulled the changes @jchodera made in to ChemicalEnvironments in #86 into this PR. I had pulled that code into my branch after writing issue #100. I haven't addressed issue #100 yet, if we just want to replace things with snake case I should be able to do that fairly easily with find and replace and could add it to this PR. When I wrote that issue I was imagining changing the `Get` and `Set` functions to also be more pythonic with `@property` and `setter` methods instead. I don't think this will take me very long, but its been REALY low on the priority list with SAMPL and MolSSI deadlines approaching. 

So currently this PR addresses only issue #106 which involved more than one bug that I found:

1. aromatic atom symbols weren't recognized when looking for atoms not in brackets. For example, `'c1ccccc1'` is technically a SMIRKS string, but before ChemicalEnvironments would have complained because it didn't know that `'c'` was an atom. Now it does. 

2. You can specify rings in recursive SMIRKS such as `"[#6X3$(*1-C=C-C1):1]"` could be used to look for a butene like group. However, the parsing didn't put the first `1` after the `*` in the right place before. 

3. While addressing 2, I discovered another bug with recursive SMIRKS. The first atom inside the recursion always refers to the first or "top" atom. In my mind, there are 3 options for what that atom could look like. 
    1.  `*` which means there is no additional information `"[#6X3$(*~[#7,#8,#9,#17,#35,#53]):1]"`. These are the most common version of this type of SMIRKS pattern
    2. a "square bracketed atom" such as `"[$(#6X3~[#7,#8,#9,#17,#35,#53])]"`. These aren't that common in human written SMIRKS, but they can appear if you had a replacement string such as `"$ewg"` where you did the replacement before making the `ChemicalEnvironment`
    3. Atomic or aromatic symbols that do not require brackets which look like: `"[$(c1ccccc1)]"`. This last type was not handled at all in the parsing, I've added coverage for this scenario. 

I've tried to add tests to cover all three of these cases. However, I would like to take this bug as an example of reasons why just "can we put it into a `ChemicalEnvironment`" is not a sufficient test. Many of these are caught because of type mismatches or because after pulling the SMIRKS into a `ChemicalEnvironment` we check that SMIRKS created by that environment is actually valid in some other toolkit. I could update the `isValid` function to check for openeye or rdkit and use which ever is available. That function is the only place `ChemicalEnvironments` are dependent on OpenEye. 

@jchodera or @davidlmobley take a look at this description and let me know if you want me to change things to snake case or address the isValid issue before we merge this.

### CODE REVIEW

**Comment by davidlmobley** [comment]
@bannanc 
> I could update the isValid function to check for openeye or rdkit and use which ever is available

Thanks for your work on this. I think this would be useful if you can get the RDKit version of it working easily, though if not it would make sense to me to do that in a separate issue/PR later. 

I'll have a quick look through your other changes, but I'm fine having the rest of the issues be dealt with separately/later.

File: openforcefield/typing/chemistry/environment.py [hunk 1]
```diff
@@ -18,6 +18,9 @@
 and David Mobley, UC Irvine.
 
 """
+
+# QUESTION: should we ditch camelCase and go with PEP8 standard snake_case?
```
Comment (davidlmobley) [inline openforcefield/typing/chemistry/environment.py]: SGTM to do eventually; ambivalent about whether it is now.

Comment (bannanc) [reply openforcefield/typing/chemistry/environment.py]: This is a carry over from @jchodera since I took the chemenv from the Topology branch. I think the answer to this is yes eventually, but not happening right now.

**Comment by bannanc** [comment]
@davidlmobley 
Otherwise I'm happy to wait on the other pythonic issues for future PRs and merge this. I'm not allowed to merge this because the tests are failing so will you merge it unless. 

Did you want me to change this question in the python file?

> # QUESTION: should we ditch camelCase and go with PEP8 standard snake_case?

**Comment by davidlmobley** [comment]
Happy to merge, but what about:
> Thanks for your work on this. I think this would be useful if you can get the RDKit version of it working easily, though if not it would make sense to me to do that in a separate issue/PR later.

Also can you remove "WIP"?

**Comment by davidlmobley** [comment]
Mind also disabling the py27 test at the same time, since we're going to keep having problems with that one given that py27 support is going away?

**Comment by bannanc** [comment]
@davidlmobley I tried removing python 2.7 and then the other tests failed... I don't understand why, but that was [this commit](https://travis-ci.org/openforcefield/openforcefield/builds/372318575?utm_source=github_status&utm_medium=notification)

I then switched to the conda install for openeye because I thought it made sense for us to switch to that installation. (it is still failing)

**Comment by bannanc** [comment]
I found a discrepancy in openeye and rdkit SMARTS parsing, I updated the test to avoid the corner case and wrote an issue in RDKit's issue tracker: https://github.com/rdkit/rdkit/issues/1842

**Comment by bannanc** [comment]
@davidlmobley This is ready for you to take a look at again, I'm happy to handle the rdkit part differently, but I wanted to make sure I could test it so making it an init option seemed like the easiest way to do that for now.

**Comment by davidlmobley** [comment]
Looks good to me! Thanks! Update docs in a separate PR to note that py27 may still work but is no longer supported?
